### PR TITLE
BUG: DataFrame from dict with non-nano Timedelta

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

xref #44504

This is motivated by a branch (upcoming PR) that makes the Timedelta constructor preserve resolution when passed a np.timedelta64 object.  Without this change, test_constructor_dict_timedelta64_index fails because of a dict lookup in lib.fast_multiget.

For non-nano Timedeltas this does significantly hurt performance:

```
In [1]: import pandas as pd
In [2]: td = pd.Timedelta(days=1, minutes=3)
In [3]: td2 = td._as_unit("ms")

In [4]: %timeit hash(td)
145 ns ± 15.5 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)  # <- main
131 ns ± 3.43 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)  # <- PR

In [5]: %timeit hash(td2)
138 ns ± 16.8 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)  # <- main
1.8 µs ± 42.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)  # <- PR
```

I expect there's room for optimization.  Something like:

```
if self._is_in_pytimedelta_bounds() and not self._has_ns():
    return timedelta.__hash__(self)
else:
    return self._hash_version_in this_pr()
```

### CODE REVIEW

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 1]
```diff
@@ -1082,8 +1082,19 @@ cdef class _Timedelta(timedelta):
...
-        else:
+        elif self._reso == NPY_FR_ns:
             return timedelta.__hash__(self)
+        else:
+            # We want to ensure that two equivalent Timedelta objects
+            #  have the same hash.  So we try downcasting to the next-lowest
+            #  resolution.
+            try:
+                obj = self._as_reso(<NPY_DATETIMEUNIT>(self._reso + 1))
+            except OverflowError:
```
Comment (mroeschke) [inline pandas/_libs/tslibs/timedeltas.pyx:1093]: Do the unit test hit both the `except` and `else` branch?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/timedeltas.pyx:1093]: just reach one branch, will add test(s)

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 2]
```diff
@@ -1152,6 +1170,13 @@ cdef class _Timedelta(timedelta):
         else:
             raise NotImplementedError(self._reso)
 
+    cdef bint _is_in_pytimedelta_bounds(self):
+        """
+        Check if we are within the bounds of datetime.timedelta.
+        """
+        self._ensure_components()
+        return -999999999 <= self._d and self._d <= 999999999
```
Comment (mroeschke) [inline pandas/_libs/tslibs/timedeltas.pyx:1178]: Worth using `timedelta.max.days`?

Comment (mroeschke) [reply pandas/_libs/tslibs/timedeltas.pyx:1178]: Also why doesn't the seconds and microseconds need to be checked?

File: pandas/_libs/tslibs/timedeltas.pyx (hunk 1 above)
Comment (mroeschke) [reply pandas/_libs/tslibs/timedeltas.pyx:1111]: Did these get added?

File: pandas/_libs/tslibs/timedeltas.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timedeltas.pyx:1178]: i hard-coded the 9s for perf.  seconds/microseconds are bounded so will always be in range

File: pandas/_libs/tslibs/timedeltas.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timedeltas.pyx:1111]: good catch, not yet

Comment (jbrockmendel) [reply pandas/_libs/tslibs/timedeltas.pyx:1112]: updated per request

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**